File: src/004.StackAndRecursion/020.ValidParentheses/20.有效的括号.js
/*
 * @lc app=leetcode.cn id=20 lang=javascript
 *
 * [20] 有效的括号
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
	const len = s.length;

	if (len % 2 === 1) return false;

	const pairs = new Map([
		['{', '}'],
		['(', ')'],
		['[', ']'],
	]);
	let stack = [];

	for (const v of s) {
		if (pairs.has(v)) {
			// 左括号
			stack.push(v);
		} else {
			// 右括号
			if (!stack.length || v != pairs.get(stack[stack.length - 1]))
				return false;
			stack.pop();
		}
	}

	return !stack.length;
};
// console.log(isValid('()'));
// @lc code=end

File: src/004.StackAndRecursion/020.ValidParentheses/20.有效的括号.go
/*
 * @lc app=leetcode.cn id=20 lang=golang
 *
 * [20] 有效的括号
 */

// @lc code=start
func isValid(s string) bool {
	n := len(s)

	if n%2 == 1 {
		return false
	}

	pairs := map[byte]byte{
		'(': ')',
		'{': '}',
		'[': ']',
	}
	stacks := []byte{}

	for i := 0; i < n; i++ {
		if pairs[s[i]] > 0 {
			// 左括号
			stacks = append(stacks, s[i])
		} else {
			// 右括号
			if len(stacks) == 0 || s[i] != pairs[stacks[len(stacks)-1]] {
				return false
			}
			stacks = stacks[:len(stacks)-1]
		}
	}

	return len(stacks) == 0
}

// @lc code=end

